File: sorting.py
"""
СОРТИРОВКИ
ВВОД: числа через пробел
"""
import random


def selection_sort(main_list):
    for i in range(len(main_list) - 1):
        min_index = i
        for j in range(i + 1, len(main_list)):
            if main_list[j] < main_list[min_index]:
                min_index = j
        main_list[i], main_list[min_index] = main_list[min_index], main_list[i]


def insertion_sort(main_list):
    for i in range(1, len(main_list)):
        for j in range(i - 1, - 1, -1):
            if main_list[i] > main_list[j]:
                break
            main_list[i], main_list[j] = main_list[j], main_list[i]
            i -= 1


def bubble_sort(main_list):
    for i in range(len(main_list) - 1, 0, -1):
        for j in range(i):
            if main_list[j] > main_list[j + 1]:
                main_list[j], main_list[j + 1] = main_list[j + 1], main_list[j]


def merge_sort(main_list):

    def merge(list_1, list_2):
        i = 0
        j = 0
        k = 0
        res_list = list()
        while i < len(list_1) and j < len(list_2):
            if list_1[i] <= list_2[j]:
                res_list.append(list_1[i])
                i += 1
            else:
                res_list.append(list_2[j])
                j += 1
            k += 1
        if i < len(list_1):
            res_list.append(list_1[i:])
            return res_list
        res_list.append(list_2[j:])
        return res_list

    if len(main_list) <= 1:
        return main_list
    m = len(main_list) // 2
    return merge(merge_sort(main_list[0:m + 1]), merge_sort(main_list[m + 1:]))


def quick_sort(main_list, l=None, r=None):

    def partition(main_list, l, r):

        # тут берем случайный индекс, чтоб не тормозить на почти упорядовенных массивах
        random_index = random.randint(l, r)
        main_list[l], main_list[random_index] = main_list[random_index], main_list[l]

        nowItem = main_list[l]
        j = l
        for i in range(l + 1, r + 1):
            if main_list[i] <= nowItem:
                j += 1
                main_list[j], main_list[i] = main_list[i], main_list[j]
        main_list[l], main_list[j] = main_list[j], main_list[l]
        return j

    l = l or 0
    r = r or len(main_list) - 1

    """
    Вариант с хвостовой рекурсией
    if l >= r:
        return
    m = partition(main_list, l, r)
    quick_sort(main_list, l, m - 1)
    quick_sort(main_list, m + 1, r)
    """

    # Элиминация хвостовой рекурсии и балансировка дерева(рекурсивно вызываем более короткий отрезок)
    while l < r:
        m = partition(main_list, l, r)
        if (r - m) >= m + 1:
            quick_sort(main_list, l, m - 1)
            l = m + 1
            continue
        quick_sort(main_list, m + 1, r)
        r = m - 1


def heap_sort(main_list):

    def sift_down(main_list, index, size):

        now_index = index

        while now_index < size and now_index * 2 <= size:
            target_index = now_index * 2
            if (now_index * 2) + 1 <= size and \
                    main_list[(now_index * 2) - 1] < main_list[((now_index * 2) + 1) - 1]:
                target_index = (now_index * 2) + 1
            if main_list[now_index - 1] >= main_list[target_index - 1]:
                break
            main_list[now_index - 1], main_list[target_index - 1] = \
                main_list[target_index - 1], main_list[now_index - 1]
            now_index = target_index

    def build_max_heap(main_list):
        for i in range(len(main_list) // 2, 0, -1):
            sift_down(main_list, i, len(main_list))

    build_max_heap(main_list)
    size = len(main_list)
    for i in range(len(main_list), 1, - 1):
        main_list[size - 1], main_list[0] = main_list[0], main_list[size - 1]
        size -= 1
        sift_down(main_list, 1, size)



if __name__ == '__main__':
    sort_list = [int(i) for i in input().split()]
    heap_sort(sort_list)
    print(*sort_list)

File: Sorting.java
import java.util.Scanner;
/*
СОРТИРОВКИ
ВВОД: первое число - кол-во элтов в массиве, далее - числа в массиве через пробел
 */
class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int[] list = new int[n];
        for (int i = 0; i < n; i++) {
            list[i] = scanner.nextInt();
        }
        heapSort(list);
        StringBuilder out = new StringBuilder();
        for (int i =0; i < n; i++) {
            out.append(list[i]);
            out.append(" ");
        }
        System.out.println(out);
    }

    public static void selectionSort(int[] list) {
        for (int i = 0 ; i < list.length - 1; i++) {
            int indexOfMin = i;
            for (int j = i + 1; j < list.length; j++) {
                if (list[j] < list[indexOfMin]) {
                    indexOfMin = j;
                }
            }
            int temp = list[i];
            list[i] = list[indexOfMin];
            list[indexOfMin] = temp;
        }
    }

    public static void insertSort(int[] list) {
        for (int i = 1; i < list.length; i++) {
            int j = i;
            while (j > 0 && list[j] < list[j - 1]) {
                int temp = list[j];
                list[j] = list[j - 1];
                list[j - 1] = temp;
                j -= 1;
            }
        }
    }

    public static void bubbleSort(int[] list) {
        for (int i = list.length - 1; i > 0; i--) {
            int j = 0;
            while (j < i) {
                if (list[j] > list[j + 1]) {
                        int temp = list[j];
                        list[j] = list[j + 1];
                        list[j + 1] = temp;
                }
                j += 1;
            }
        }
    }

    public static int[] merge(int[] a, int[] b) {
        int i = 0;
        int j = 0;
        int k = 0;
        int[] res = new int[a.length + b.length];
        while (i < a.length && j < b.length) {
            if (a[i] <= b[j]) {
                res[k] = a[i];
                i += 1;
            } else {
                res[k] = b[j];
                j += 1;
            }
            k += 1;
        }
        if (i < a.length) {
            while (i < a.length) {
                res[k] = a[i];
                i +=1;
            }
        } else {
            while (j < b.length) {
            res[k] = b[j];
            j += 1;
            }
        }
        return res;
    }

    public static int[] mergeSort(int[]list) {
        if (list.length <= 1) {
            return list;
        }
        int m = list.length / 2;
        int[] list1 = new int[m];
        int[] list2 = new int[list.length - m];
        for (int i= 0; i < list.length; i++){
            if (i < m) {
                list1[i] = list[i];
            } else {
                list2[i - m] = list[i];
            }
        }
        return merge(mergeSort(list1), mergeSort(list2));
    }


    public static int partition(int[] list, int l, int r) {
        int j = l;
        for (int i = l + 1; i <= r; i++) {
            if (list[i] <= list[l]) {
                j += 1;
                int temp = list[j];
                list[j] = list[i];
                list[i] = temp;
            }
        }
        int temp = list[j];
        list[j] = list[l];
        list[l] = temp;
        return j;
    }

    public static void quickSort(int[] list, int l, int r) {
        if (l >= r) {
            return;
        }
        int m = partition(list, l, r);
        quickSort(list, l, m - 1);
        quickSort(list, m + 1, r);
    }

    public static void quickSort(int[] list) {
        quickSort(list, 0, list.length - 1);
    }

    public static void siftDownMaxHeap(int[] list, int index, int size) {
        int nowIndex = index;
        while (nowIndex < size && (nowIndex * 2) <= size) {
            int targetIndex = nowIndex * 2;
            if ((nowIndex * 2) + 1 <= size &&
                    list[(nowIndex * 2) - 1] < list[((nowIndex * 2) + 1) - 1]) {
                targetIndex = (nowIndex * 2) + 1;
            }
            if (list[nowIndex - 1] > list[targetIndex - 1]) {
                break;
            }
            int temp = list[nowIndex - 1];
            list[nowIndex - 1] = list[targetIndex - 1];
            list[targetIndex - 1] = temp;
            nowIndex = targetIndex;
        }
    }

    public static void buildMaxHeap(int[] list) {
        for (int i = (list.length / 2); i > 0; i--) {
            siftDownMaxHeap(list, i, list.length);
        }
    }

    public static void heapSort(int[] list) {
        buildMaxHeap(list);
        int size = list.length;
        for (int i = list.length; i > 0; i--) {
            int temp = list[size - 1];
            list[size - 1] = list[0];
            list[0] = temp;
            size -= 1;
            siftDownMaxHeap(list, 1, size);
        }
    }
}
